File: main_init_mysql_to_sqlserver.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"os"
	"time"

	mssql "github.com/denisenkom/go-mssqldb"
	_ "github.com/go-sql-driver/mysql"
	"github.com/joho/godotenv"
)

// Fungsi menentukan tanggal_shift seperti di Python
func getShiftDate(t time.Time) *time.Time {
	h := t.Hour()
	if h >= 7 && h <= 18 {
		d := time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 1, 0, t.Location())
		return &d
	} else if h >= 19 && h <= 23 {
		d := time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 2, 0, t.Location())
		return &d
	} else {
		prev := t.AddDate(0, 0, -1)
		d := time.Date(prev.Year(), prev.Month(), prev.Day(), 0, 0, 2, 0, t.Location())
		return &d
	}
}

func normalizeValue(v interface{}) interface{} {
	switch val := v.(type) {
	case nil:
		return nil
	case []byte:
		return string(val)
	default:
		return val
	}
}

func main() {
	fmt.Println("=== 🚀 Initial Sync MySQL → SQL Server (Bulk Mode) ===")

	startTime := time.Now()
	fmt.Printf("🕒 Mulai pada: %s\n", startTime.Format("2006-01-02 15:04:05"))

	err := godotenv.Load()
	if err != nil {
		log.Fatalf("Error loading .env: %v", err)
	}

	// --- Konfigurasi ---
	mysqlHost := os.Getenv("MYSQL_HOST")
	mysqlUser := os.Getenv("MYSQL_USER")
	mysqlPass := os.Getenv("MYSQL_PASS")
	mysqlDB := os.Getenv("MYSQL_DB")

	sqlsrvHost := os.Getenv("SQLSERVER_HOST")
	sqlsrvDB := os.Getenv("SQLSERVER_DB")
	sqlsrvUser := os.Getenv("SQLSERVER_USER")
	sqlsrvPass := os.Getenv("SQLSERVER_PASS")

	mysqlTable := os.Getenv("MYSQL_TABLE")
	sqlsrvTable := os.Getenv("SQLSRV_TABLE")
	wbTag := os.Getenv("WB_TAG")
	if wbTag == "" {
		wbTag = "DEFAULT_WB"
	}

	// --- Koneksi database ---
	mysqlDSN := fmt.Sprintf("%s:%s@tcp(%s)/%s?parseTime=true", mysqlUser, mysqlPass, mysqlHost, mysqlDB)
	mysqlConn, err := sql.Open("mysql", mysqlDSN)
	if err != nil {
		log.Fatalf("Koneksi MySQL gagal: %v", err)
	}
	defer mysqlConn.Close()

	sqlsrvDSN := fmt.Sprintf("sqlserver://%s:%s@%s?database=%s", sqlsrvUser, sqlsrvPass, sqlsrvHost, sqlsrvDB)
	sqlsrvConn, err := sql.Open("sqlserver", sqlsrvDSN)
	if err != nil {
		log.Fatalf("Koneksi SQL Server gagal: %v", err)
	}
	defer sqlsrvConn.Close()

	fmt.Println("=== [Initial Sync] Pemeriksaan awal... ===")

	// --- Cek tabel SQL Server kosong ---
	var count int
	err = sqlsrvConn.QueryRow(fmt.Sprintf("SELECT COUNT(*) FROM %s", sqlsrvTable)).Scan(&count)
	if err != nil {
		log.Fatalf("Gagal menghitung data SQL Server: %v", err)
	}
	if count > 0 {
		fmt.Printf("❌ Inisialisasi dibatalkan: Tabel %s sudah berisi %d data.\n", sqlsrvTable, count)
		return
	}

	// --- Ambil semua data MySQL ---
	rows, err := mysqlConn.Query(fmt.Sprintf("SELECT * FROM %s", mysqlTable))
	if err != nil {
		log.Fatalf("Gagal ambil data MySQL: %v", err)
	}
	defer rows.Close()

	mysqlCols, _ := rows.Columns()
	allCols := append(mysqlCols, "TANGGAL_SHIFT", "DATE_SYNC", "WB_TAG", "DELETED")

	fmt.Printf("Menyalin data dari %s → %s menggunakan Bulk Insert...\n", mysqlTable, sqlsrvTable)

	tx, err := sqlsrvConn.Begin()
	if err != nil {
		log.Fatalf("Gagal mulai transaksi SQL Server: %v", err)
	}

	stmt, err := tx.Prepare(mssql.CopyIn(sqlsrvTable, mssql.BulkOptions{}, allCols...))
	if err != nil {
		log.Fatalf("Gagal siapkan bulk insert: %v", err)
	}

	totalInserted := 0
	mysqlVals := make([]interface{}, len(mysqlCols))
	mysqlPtrs := make([]interface{}, len(mysqlCols))
	for i := range mysqlVals {
		mysqlPtrs[i] = &mysqlVals[i]
	}

	for rows.Next() {
		if err := rows.Scan(mysqlPtrs...); err != nil {
			log.Printf("⚠️ Gagal membaca row: %v", err)
			continue
		}

		insertVals := make([]interface{}, 0, len(allCols))
		for i := range mysqlVals {
			insertVals = append(insertVals, normalizeValue(mysqlVals[i]))
		}

		// Hitung tanggal_shift
		var tanggal2 time.Time
		for i, col := range mysqlCols {
			if col == "TANGGAL2" {
				if val, ok := mysqlVals[i].(time.Time); ok {
					tanggal2 = val
				}
				break
			}
		}

		var tanggalShift *time.Time
		if !tanggal2.IsZero() {
			tanggalShift = getShiftDate(tanggal2)
		}

		insertVals = append(insertVals, tanggalShift)
		insertVals = append(insertVals, time.Now())
		insertVals = append(insertVals, wbTag)
		insertVals = append(insertVals, 0)

		_, err = stmt.Exec(insertVals...)
		if err != nil {
			log.Printf("⚠️ Gagal bulk insert row: %v", err)
			continue
		}

		totalInserted++
		if totalInserted%10000 == 0 {
			fmt.Printf("Progress: %d baris...\n", totalInserted)
		}
	}

	// Tutup Bulk Insert
	_, err = stmt.Exec()
	if err != nil {
		log.Fatalf("Gagal eksekusi akhir bulk insert: %v", err)
	}
	if err := stmt.Close(); err != nil {
		log.Fatalf("Gagal menutup bulk statement: %v", err)
	}
	if err := tx.Commit(); err != nil {
		log.Fatalf("Gagal commit bulk insert: %v", err)
	}

	endTime := time.Now()
	fmt.Printf("✅ Selesai. Total %d baris disalin.\n", totalInserted)
	fmt.Printf("🕔 Selesai pada: %s\n", endTime.Format("2006-01-02 15:04:05"))
	fmt.Printf("⏱️ Durasi total: %s\n", endTime.Sub(startTime).Round(time.Second))
	fmt.Println("=== Proses selesai ===")
}

File: main_init_mysql_to_sqlserver.py
import os
from dotenv import load_dotenv
import time
import datetime
import mysql.connector
import pyodbc

load_dotenv()

MYSQL_CONN = {
    'host': os.getenv("MYSQL_HOST"),
    'user': os.getenv("MYSQL_USER"),
    'password': os.getenv("MYSQL_PASS"),
    'database': os.getenv("MYSQL_DB")
}

SQLSERVER_CONN = {
    'server': os.getenv("SQLSERVER_HOST"),
    'database': os.getenv("SQLSERVER_DB"),
    'username': os.getenv("SQLSERVER_USER"),
    'password': os.getenv("SQLSERVER_PASS")
}

MYSQL_TABLE = os.getenv("MYSQL_TABLE")
SQLSRV_TABLE = os.getenv("SQLSRV_TABLE")
WB_TAG = os.getenv("WB_TAG")

def get_shift_date(dt):
    if not dt:
        return None
    try:
        trim_date = dt.date()
        trim_time = dt.time()
        if datetime.time(7, 0, 0) <= trim_time <= datetime.time(18, 59, 59):
            return datetime.datetime.combine(trim_date, datetime.time(0, 0, 1))
        elif datetime.time(19, 0, 0) <= trim_time <= datetime.time(23, 59, 59):
            return datetime.datetime.combine(trim_date, datetime.time(0, 0, 2))
        else:
            return datetime.datetime.combine(trim_date - datetime.timedelta(days=1), datetime.time(0, 0, 2))
    except Exception as e:
        print("Error get_shift_date:", e)
        return None

def initial_sync():
    print("=== [Initial Sync] Pemeriksaan awal... ===")

    sqlsrv_conn = pyodbc.connect(
        f"DRIVER={{ODBC Driver 17 for SQL Server}};"
        f"SERVER={SQLSERVER_CONN['server']};"
        f"DATABASE={SQLSERVER_CONN['database']};"
        f"UID={SQLSERVER_CONN['username']};"
        f"PWD={SQLSERVER_CONN['password']}",
        autocommit=False
    )
    sqlsrv_cur = sqlsrv_conn.cursor()
    sqlsrv_cur.fast_executemany = True  # 🚀 aktifkan mode cepat

    mysql_conn = mysql.connector.connect(**MYSQL_CONN)
    mysql_cur = mysql_conn.cursor(dictionary=True)

    try:
        sqlsrv_cur.execute(f"SELECT COUNT(*) FROM {SQLSRV_TABLE} WHERE wb_tag = ?", (WB_TAG,))
        count_sqlsrv = sqlsrv_cur.fetchone()[0]

        if count_sqlsrv > 0:
            print(f"❌ Data dengan WB_TAG={WB_TAG} sudah ada di tabel {SQLSRV_TABLE} ({count_sqlsrv} baris).")
            return

        mysql_cur.execute(f"SELECT * FROM {MYSQL_TABLE}")
        rows = mysql_cur.fetchall()
        total_rows = len(rows)
        if total_rows == 0:
            print("⚠️ Tidak ada data untuk disalin.")
            return

        print(f"Menyalin {total_rows} baris dari MySQL ke SQL Server...")
        
        batch_data = []
        for row in rows:
            tanggal_shift = get_shift_date(row.get("TANGGAL2"))
            row["tanggal_shift"] = tanggal_shift
            row["date_sync"] = datetime.datetime.now()
            row["wb_tag"] = WB_TAG
            row["deleted"] = 0
            batch_data.append(row)

        col_names = list(batch_data[0].keys())
        col_list_sql = ", ".join(f"[{c}]" for c in col_names)
        placeholders = ", ".join("?" for _ in col_names)
        insert_sql = f"INSERT INTO {SQLSRV_TABLE} ({col_list_sql}) VALUES ({placeholders})"

        params_list = [[row[c] for c in col_names] for row in batch_data]
        sqlsrv_cur.executemany(insert_sql, params_list)
        sqlsrv_conn.commit()

        print(f"✅ Selesai! Total {total_rows} baris disalin ke SQL Server.")
    except Exception as e:
        print(f"❌ Error fatal: {e}")
        sqlsrv_conn.rollback()
    finally:
        mysql_cur.close()
        mysql_conn.close()
        sqlsrv_cur.close()
        sqlsrv_conn.close()

if __name__ == "__main__":
    try:
        start = time.time()
        start_time = datetime.datetime.now()

        print("=== Program Initial Sync MySQL → SQL Server ===")
        print(f"🕒 Mulai pada: {start_time.strftime('%Y-%m-%d %H:%M:%S')}")
        
        initial_sync()
        
        durasi = round(time.time() - start)
        selesai_pada = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        
        print(f"🕔 Selesai pada: {selesai_pada}")
        print(f"⏱️ Durasi total: {durasi}s")
    
    except Exception as e:
        print("\n‼️ Terjadi error:")
        print(e)

    finally:
        input("\nTekan Enter untuk keluar...")